impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -131,79 +147,55 @@ private KafkaConfig buildBrokerConfig(KafkaClusterConfig.ConfigHolder c, Path te
-            }
-
-            @Override
-            public EndpointPair getInterBrokerEndpoint(int brokerId) {
-                return buildEndpointPair(interBrokerPorts, brokerId);
-            }
-
-            @Override
-            public EndpointPair getControllerEndpoint(int brokerId) {
-                return buildEndpointPair(controllerPorts, brokerId);
-            }
-
-            private EndpointPair buildEndpointPair(List<ServerSocket> portRange, int brokerId) {
-                var port = portRange.get(brokerId);
-                return EndpointPair.builder().bind(new Endpoint("0.0.0.0", port.getLocalPort())).connect(new Endpoint("localhost", port.getLocalPort())).build();
-            }
-        };
-
         buildAndStartZookeeper();
-        servers = clusterConfig.getBrokerConfigs(() -> kafkaEndpoints).parallel().map(configHolder -> {
+        clusterConfig.getBrokerConfigs(() -> this).parallel().forEach(configHolder -> {
             final Server server = this.buildKafkaServer(configHolder);
-            Utils.awaitCondition(STARTUP_TIMEOUT, TimeUnit.SECONDS)
-                    .until(() -> {
-                        // Hopefully we can remove this once a fix for https://issues.apache.org/jira/browse/KAFKA-14908 actually lands.
-                        try {
-                            server.startup();
-                            return true;
-                        }
-                        catch (Throwable t) {
-                            LOGGER.log(System.Logger.Level.WARNING, "failed to start server due to: " + t.getMessage());
-                            LOGGER.log(System.Logger.Level.WARNING, "anon: {0}, client: {1}, controller: {2}, interBroker: {3}, ",
-                                    kafkaEndpoints.getAnonEndpoint(configHolder.getBrokerNum()).getBind(),
-                                    kafkaEndpoints.getClientEndpoint(configHolder.getBrokerNum()).getBind(),
-                                    kafkaEndpoints.getControllerEndpoint(configHolder.getBrokerNum()).getBind(),
-                                    kafkaEndpoints.getInterBrokerEndpoint(configHolder.getBrokerNum()).getBind());
-
-                            server.shutdown();
-                            server.awaitShutdown();
-                            return false;
-                        }
-                    });
-            return server;
-        })
-                .collect(Collectors.toList());
-
-        Utils.awaitExpectedBrokerCountInClusterViaTopic(clusterConfig.getAnonConnectConfigForCluster(kafkaEndpoints), 120, TimeUnit.SECONDS,
+            tryToStartServerWithRetry(configHolder, server);
+            servers.put(configHolder.getBrokerNum(), server);
+
+        });
+        Utils.awaitExpectedBrokerCountInClusterViaTopic(
+                clusterConfig.getConnectConfigForCluster(buildServerList(brokerId -> getEndpointPair(Listener.ANON, brokerId)), null, null, null, null), 120,
```

Comment:
Why did you get rid of the wrapper methods which hid this forest of `null` args?

Reply:
that was unintentional, I'll fix that.